src/main/java/org/openjdk/shenandoah/ShenandoahVisualizer.java

```diff
@@ -141,250 +429,878 @@ public void componentResized(ComponentEvent ev) {
             c.weightx = 1;
             c.weighty = 1;
             c.insets = pad;
+            c.gridheight = GridBagConstraints.REMAINDER;
             frame.add(legendPanel, c);
         }
 
-        frame.setVisible(true);
+        toolbarPanel.addKeyListener(keyShortcutAdapter);
+        toolbarPanel.setFocusable(true);
+        toolbarPanel.requestFocusInWindow();
 
-        ScheduledExecutorService service = Executors.newScheduledThreadPool(1);
-        ScheduledFuture<?> f = service.scheduleAtFixedRate(render,
-                0, 100, TimeUnit.MILLISECONDS);
+        frame.setVisible(true);
 
         frame.addWindowListener(new WindowAdapter() {
             public void windowClosing(WindowEvent e) {
+                f[0].cancel(true);
                 service.shutdown();
                 frame.dispose();
             }
         });
+        f[0].get();
```

Comment:
Regarding the refactoring again, it seems odd that you have a final local array field here, only to allow accessing the first element. I feel that this should go into a separate class, and a lot of other stuff too.